ktor-server/jvm/src/io/ktor/server/application/plugins/api/ServerPlugin.kt

```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright 2014-2020 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+@file:Suppress("DEPRECATION")
+
+package io.ktor.server.application.plugins.api
+
+import io.ktor.server.application.*
+import io.ktor.server.config.*
+import io.ktor.server.request.*
+import io.ktor.server.response.*
+import io.ktor.util.*
+import io.ktor.util.pipeline.*
+import kotlin.random.Random
+
+/**
+ * A plugin for Ktor that embeds into the HTTP pipeline and extends functionality of Ktor framework.
+ **/
+public abstract class ServerPlugin<Configuration : Any> private constructor(
+    installablePlugin: ServerPluginFactory<Configuration>
+) : PluginContext {
+    protected abstract val pipeline: ApplicationCallPipeline
+
+    public abstract val pluginConfig: Configuration
+
+    public val name: String = installablePlugin.name
+
+    public val environment: ApplicationEnvironment? get() = pipeline.environment
+    public val configuration: ApplicationConfig? get() = environment?.config
+
+    private val key: AttributeKey<ServerPlugin<Configuration>> = installablePlugin.key
+
+    internal val callInterceptions: MutableList<CallInterception> = mutableListOf()
+
+    internal val monitoringInterceptions: MutableList<CallInterception> = mutableListOf()
```

Comment:
do we still need monitoring? Please check others as well

Reply:
No. This interceptions are not used in public Api, will remove